trin-core/src/utp/stream.rs

```diff
@@ -488,179 +496,249 @@ impl UtpSocket {
+                Err(anyhow!("Connection reset by remote peer"))
+            }
+            (state, ty) => {
+                let message = format!("Unimplemented handling for ({state:?},{ty:?})");
+                debug!("{}", message);
+                Err(anyhow!(message))
+            }
         }
     }
 
-    fn handle_data_packet(&mut self, packet: Packet) {
-        if self.state == SocketState::SynRecv {
-            self.state = SocketState::Connected;
-        }
+    fn prepare_reply(&self, original: &Packet, t: PacketType) -> Packet {
+        let mut resp = Packet::new();
+        resp.set_type(t);
+        let self_t_micro = now_microseconds();
+        let other_t_micro = original.timestamp();
+        let time_difference: Delay = abs_diff(self_t_micro, other_t_micro);
+        resp.set_timestamp(self_t_micro);
+        resp.set_timestamp_difference(time_difference);
+        resp.set_connection_id(self.sender_connection_id);
+        resp.set_seq_nr(self.seq_nr);
+        resp.set_ack_nr(self.ack_nr);
+
+        resp
+    }
 
-        let mut packet_reply = Packet::new();
-        packet_reply.set_type(PacketType::State);
-        packet_reply.set_connection_id(self.sender_connection_id);
-        packet_reply.set_seq_nr(self.seq_nr);
-        packet_reply.set_ack_nr(self.ack_nr);
+    fn handle_data_packet(&mut self, packet: &Packet) -> Option<Packet> {
+        // If a FIN was previously sent, reply with a FIN packet acknowledging the received packet.
+        let packet_type = if self.state == SocketState::FinSent {
+            PacketType::Fin
+        } else {
+            PacketType::State
+        };
+        let mut reply = self.prepare_reply(packet, packet_type);
 
         if packet.seq_nr().wrapping_sub(self.ack_nr) > 1 {
-            let sack_bitfield = self.build_selective_ack();
-            packet_reply.set_selective_ack(sack_bitfield);
-        }
-
-        self.send_packet(&mut packet_reply);
-
-        // Add packet to BTreeMap
-        self.incoming_buffer.insert(packet.seq_nr(), packet);
+            debug!(
+                "current ack_nr ({}) is behind received packet seq_nr ({})",
+                self.ack_nr,
+                packet.seq_nr()
+            );
 
-        // TODO: use pop_front when it is in a stable release
-        if let Some(packet_from_buffer) = self.incoming_buffer.clone().values().next() {
-            let packet_seq = packet_from_buffer.seq_nr();
-            if !self.incoming_buffer.is_empty()
-                && (self.ack_nr == packet_seq || self.ack_nr + 1 == packet_seq)
-            {
-                self.incoming_buffer.remove(&packet_seq);
-                self.ack_nr = packet_seq;
+            // Set SACK extension payload if the packet is not in order
+            let sack_bitfield = self.build_selective_ack();
 
-                self.recv_data_stream
-                    .append(&mut Vec::from(packet_from_buffer.payload()));
+            if !sack_bitfield.is_empty() {
```

Comment:
*nitpick* but I tend to prefer `is_some()` over `!...is_empty()`

Reply:
In Rust, `is_some()` is a method implemented for `Option<T>`, no such a method for `Vec<T>`.